lib/facter/self_service.rb

```diff
@@ -114,6 +114,81 @@
     { S0014: res.nil? }
   end
 
+  chunk(:S0016) do
+    # Puppetserver
+    next unless PuppetSelfService.primary? || PuppetSelfService.compiler? || PuppetSelfService.legacy_compiler?
+    log_path = Puppet.settings['logdir'].to_s + '/../puppetserver/puppetserver.log'
+    line_count = `wc -l < #{log_path}`.to_i
+    lines_to_read = 250
+    # if the number of lines in the logfile is smaller than the amount we want to read, set to read all lines in the file
+    if line_count < lines_to_read
+      lines_to_read = line_count
+    end
+    range_from = line_count - lines_to_read
+    # read each line and match for error
+    lines = IO.readlines(log_path)[range_from..line_count]
+    result = 'dummy'
+    lines.each_index.each do |i|
+      if lines[i].include?('java.lang.OutOfMemoryError')
+        result = false
+        break
+      else
+        result = true
+      end
+    end
+    { S0016: result }
```

Comment:
We should also look for a recent (within the last `runinterval`) `<service>/<service>_err_pid*.log`. E.g `puppetserver_err_pid15518.log`. This will denote an OOM failure in the service so we can skip searching through the logs.

see `https://tickets.puppetlabs.com/browse/EZ-132` for more details.